### PR TITLE
Allows setting device values through MQTT /set endpoint

### PR DESCRIPTION
**Pull request by Meatballs1**
Resolves https://github.com/hcpy2-0/hcpy/issues/6

Also removes the attempts to shoehorn the attributes into specific topics. We should just let all the device attributes flow through as JSON and let the receiver handle making it nice if required (for now).

### CODE REVIEW

File: HCDevice.py [hunk 1]
```diff
@@ -92,19 +93,62 @@ def parse_values(self, values):
...
+		if 'access' not in feature:
+			raise Exception(f"{self.name}. Unable to configure appliance. Feature {feature['name']} with uid {uid} does not have access.")
+
+		access = feature['access'].lower()
+		if access != 'readwrite' and access != 'writeonly':
+			raise Exception(f"{self.name}. Unable to configure appliance. Feature {feature['name']} with uid {uid} has got access {feature['access']}.")
+
+		# check if selected list with values is allowed
+		if 'values' in feature:
+			if isinstance(data['value'], int) == False:
```
Comment (pmagyar) [inline HCDevice.py:125]: comparison to False should be 'if cond is False:' or 'if not cond:'
https://www.flake8rules.com/rules/E712.html

Or is there a reason for this implementation?

File: HCDevice.py [hunk 2]
```diff
@@ -92,19 +93,62 @@ def parse_values(self, values):
...
+				raise Exception(f"Unable to configure appliance. The value {data['value']} must be an integer. Allowed values are {feature['values']}.")
+			value = str(data['value']) # values are strings in the feature list, but always seem to be an integer. An integer must be provided
+			if value not in feature['values']:
+				raise Exception(f"{self.name}. Unable to configure appliance. Value {data['value']} is not a valid value. Allowed values are {feature['values']}.")
+
+		if 'min' in feature:
+			min = int(feature['min'])
+			max = int(feature['max'])
+			if isinstance(data['value'], int) == False or data['value'] < min or data['value'] > max:
```
Comment (pmagyar) [inline HCDevice.py:134]: comparison to False should be 'if cond is False:' or 'if not cond:'
https://www.flake8rules.com/rules/E712.html

Or is there a reason for this implementation?

File: HCDevice.py [hunk 3]
```diff
@@ -129,25 +173,32 @@ def get(self, resource, version=1, action="GET", data=None):
 		}
 
 		if data is not None:
-			msg["data"] = [data]
+			if action == "POST":
+				if self.test_feature(data) != True:
```
Comment (pmagyar) [inline HCDevice.py:177]: E712 comparison to True should be 'if cond is not True:' or 'if not cond:'
 
 Or is there a reason for this implementation?

File: README.md [hunk 4]
```diff
@@ -361,9 +359,70 @@ The coffee machine needs a better mapping to MQTT messages.
     'ProcessPhase': 'None'
 }
 ```
-
 </details>
 
+## Posting to the appliance
+
+Whereas the reading of de status is very beta, this is very very alpha. There is some basic error handling, but don't expect that everything will work.
```
Comment (pmagyar) [inline README.md:366]: de = the ?

File: HCDevice.py [hunk 5]
```diff
@@ -92,19 +93,62 @@ def parse_values(self, values):
 
 		return result
 
+	# Test the feature of an appliance agains a data object
+	def test_feature(self, data):
+		if 'uid' not in data:
+			raise Exception("{self.name}. Unable to configure appliance. UID is required.")
+
+		if isinstance(data['uid'], int) == False:
```
[the remark this answers is not in the document]
Comment (Meatballs1) [reply HCDevice.py:101]: No reason as far as I can see I will amend

File: hc2mqtt [hunk 6]
```diff
@@ -12,89 +12,126 @@ import paho.mqtt.client as mqtt
...
 def client_connect(client, device, mqtt_topic):
+	def on_message(client, userdata, msg):
+		print(msg.topic)
+		mqtt_state = msg.payload.decode()
+		mqtt_topic = msg.topic.split('/')
+		print(now(),f"received mqtt message {mqtt_state}")
+		try:
+			msg = json.loads(mqtt_state)
+			if 'uid' in msg:
+				dev[mqtt_topic[-2]].get("/ro/values",1,"POST",msg)
```
Comment (pmagyar) [inline hc2mqtt:74]: E231 missing whitespace after ','
https://www.flake8rules.com/rules/E712.html

File: hc2mqtt [hunk 7]
```diff
@@ -12,89 +12,126 @@ import paho.mqtt.client as mqtt
...
-
+global dev
+dev = {}
 
 def client_connect(client, device, mqtt_topic):
+	def on_message(client, userdata, msg):
+		print(msg.topic)
+		mqtt_state = msg.payload.decode()
+		mqtt_topic = msg.topic.split('/')
+		print(now(),f"received mqtt message {mqtt_state}")
```
Comment (pmagyar) [inline hc2mqtt:70]: E231 missing whitespace after ','
https://www.flake8rules.com/rules/E712.html

File: hc2mqtt [hunk 8]
```diff
@@ -12,89 +12,126 @@ import paho.mqtt.client as mqtt
...
-	client = mqtt.Client()
-	client.connect(host=mqtt_host, port=1883, keepalive=70)
+	client = mqtt.Client(mqtt_clientname)
+
+	if mqtt_username and mqtt_password:
+		client.username_pw_set(mqtt_username, mqtt_password)
+
+	if mqtt_ssl:
+		if mqtt_cafile and mqtt_certfile and mqtt_keyfile:
+			client.tls_set(ca_certs=mqtt_cafile, certfile=mqtt_certfile, keyfile=mqtt_keyfile, cert_reqs=ssl.CERT_REQUIRED)
```
Comment (pmagyar) [inline hc2mqtt:43]: F821 undefined name 'ssl'

`import ssl` missing?

File: hc2mqtt [hunk 9]
```diff
@@ -12,89 +12,126 @@ import paho.mqtt.client as mqtt
...
+		else:
+			client.tls_set(cert_reqs=ssl.CERT_NONE)
```
Comment (pmagyar) [inline hc2mqtt:45]: F821 undefined name 'ssl'

import ssl missing?

**Review by pmagyar — CHANGES_REQUESTED** [review]
Please check my comments @Meatballs1.
Thanks!

**Comment by Meatballs1** [comment]
Should be resolved @pmagyar 👍

**Pull request merged**